packages/react-graphql/src/hooks/useGraphQLDocument.ts

```diff
@@ -0,0 +1,48 @@
+import {useState, useEffect} from 'react';
+import {OperationVariables} from 'apollo-client';
+import {DocumentNode} from 'graphql-typed';
+
+import {AsyncQueryComponentType} from '..';
+
+export default function useGraphQLDocument<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  documentOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+) {
+  const [document, setDocument] = useState<DocumentNode<
+    Data,
+    Variables
+  > | null>(
+    () => (isDocumentNode(documentOrComponent) ? documentOrComponent : null),
+  );
+
+  async function loadDocument() {
+    if (!isDocumentNode(documentOrComponent)) {
+      try {
+        const resolved = await documentOrComponent.resolve();
+        setDocument(resolved);
+      } catch (error) {
+        throw Error('error loading GraphQL document');
+      }
+    }
+  }
+
+  useEffect(
+    () => {
+      if (document === null) {
+        loadDocument();
```

Comment:
Because `loadDocument` is async, it actually cannot be a callback to `useEffect`.
Since I HAVE to create a separate method for it, it just feel better if the method is not indented in this giant block.

Thought I am now wondering if my variable would be different because of this...mm.... 🤔

Reply:
>  it actually cannot be a callback

True. It can still be inlined on line 37 though, no?

>  if my scope would be different and cause problems down

Should be fine IMO

Reply:
Btw. This is not really safe (as explained here: https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies).

The reason is because `loadDocument` uses `documentOrComponent`, which comes from the outside. What if this value changes?

Reply:
When you say “this” is not safe. What is this referring to?